xenstore/xenstore.go

```diff
@@ -109,6 +112,116 @@ func xs_exists(script_name string, args []string) {
 	}
 }
 
+var max_width = 80
+
+const TAG = " = \"...\""
+const XENSTORE_ABS_PATH_MAX = 3072
+const STRING_MAX = (XENSTORE_ABS_PATH_MAX + 1024)
+
+func sanitise_value(val string) string {
+	var builder strings.Builder
+
+	for _, r := range val {
+		switch {
+		case r >= ' ' && r <= '~' && r != '\\':
+			builder.WriteRune(r)
+		case r == '\t':
+			builder.WriteString("\\t")
+		case r == '\n':
+			builder.WriteString("\\n")
+		case r == '\r':
+			builder.WriteString("\\r")
+		case r == '\\':
+			builder.WriteString("\\\\")
+		case r < '\010':
+			builder.WriteString(fmt.Sprintf("%03o", r))
+		default:
+			builder.WriteString(fmt.Sprintf("x%02x", r))
+		}
+	}
+
+	return builder.String()
+}
+
+func do_xs_ls(xs xenstoreclient.XenStoreClient, path string, depth int) {
+	result, err := xs.List(path)
+	if err != nil {
+		die("xs_ls error: %v %s", err, path)
+	}
+	for _, sub_path := range result {
+		if len(sub_path) == 0 {
+			continue
+		}
+		slash := "/"
+		if len(path) > 0 && path[len(path)-1] == '/' {
+			slash = ""
+		}
+		newPath := path + slash + sub_path
+
+		col := 0
+		for col < depth {
+			fmt.Print(" ")
+			col++
+		}
+
+		n := len(sub_path)
+		if n > (max_width - len(TAG) - col) {
+			n = (max_width - len(TAG) - col)
+		}
+		fmt.Printf(sub_path[:n])
+		col += n
+
+		if len(newPath) >= STRING_MAX {
+			fmt.Println(":")
+		} else {
+			val, err := xs.Read(newPath)
+			if err != nil {
+				fmt.Println(":")
+			} else {
+				val = sanitise_value(val)
+				if (col + len(val) + len(TAG)) > max_width {
+					n := max_width - col - len(TAG)
+					if n < 0 {
+						n = 0
+					}
+					fmt.Printf(" = \"%s...\"\n", val[:n])
+				} else {
+					fmt.Printf(" = \"%s\"\n", val)
+				}
+			}
+		}
+
+		do_xs_ls(xs, newPath, depth+1)
+	}
+}
+
+func xs_ls(script_name string, args []string) {
+	if len(args) == 1 && args[0] == "-h" {
+		die("Usage: %s [ key ... ]", script_name)
+	}
+
+	const TIOCGWINSZ = 0x5413
+	winsize, err := unix.IoctlGetWinsize(int(os.Stdout.Fd()), TIOCGWINSZ)
+	if err == nil {
```

Comment:
what if the window size can't be got? there is no `else` case if an error occurred or just say `if err != nil`

Reply:
Then the default size 80 will be used.